File: docs/manual/scala/guide/cluster/code/docs/home/scaladsl/persistence/BlogState.scala
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.home.scaladsl.persistence

//#full-example
import play.api.libs.json._

object BlogState {
  val empty = BlogState(None, published = false)

  implicit val postContentFormat = Json.format[PostContent]

  implicit val format: Format[BlogState] = Json.format[BlogState]
}

final case class BlogState(content: Option[PostContent], published: Boolean) {
  def withBody(body: String): BlogState = {
    content match {
      case Some(c) =>
        copy(content = Some(c.copy(body = body)))
      case None =>
        throw new IllegalStateException("Can't set body without content")
    }
  }

  def isEmpty: Boolean = content.isEmpty
}

final case class PostContent(title: String, body: String)
//#full-example

final case class PostSummary(postId: String, title: String)

File: docs/manual/java/guide/cluster/code/docs/home/persistence/BlogState.java
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.home.persistence;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.lightbend.lagom.serialization.Jsonable;
import java.util.Optional;

// #full-example
public final class BlogState implements Jsonable {

  public static final BlogState EMPTY = new BlogState(Optional.empty(), false);

  private final Optional<PostContent> content;
  private final boolean published;

  @JsonCreator
  public BlogState(Optional<PostContent> content, boolean published) {
    this.content = content;
    this.published = published;
  }

  public BlogState withBody(String body) {
    if (isEmpty()) throw new IllegalStateException("Can't set body without content");
    PostContent c = content.get();
    return new BlogState(Optional.of(new PostContent(c.getTitle(), body)), published);
  }

  @JsonIgnore
  public boolean isEmpty() {
    return !content.isPresent();
  }

  public Optional<PostContent> getContent() {
    return content;
  }

  public boolean isPublished() {
    return published;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;

    BlogState blogState = (BlogState) o;

    if (published != blogState.published) return false;
    return content.equals(blogState.content);
  }

  @Override
  public int hashCode() {
    int result = content.hashCode();
    result = 31 * result + (published ? 1 : 0);
    return result;
  }
}
// #full-example
